### ISSUE TITLE
MOAIAnim ping pong mode infinite loop

### ISSUE THREAD

**Issue by Kaali** [moai/moai-dev#715]
On iOS we have encountered an infinite loop with the ping pong mode in MOAIAnim. It's really hard to reproduce as it happens seemingly at random. But the bug has hit us multiple times on iOS, but only with release builds. With debug mode build the problem cannot be reproduced, which makes it quite hard to debug.

When I finally caught the infinite loop in XCode debugger (debug symbols, release build). From following the disassembly I confirmed that the infinite loop was indeed in the while-loop as explained in this forum message from 2011: http://getmoai.com/forums/beta-0-9-moaianim-setmode-moaitimer-ping-pong-causes-infinity-loop-osx-lion-t209/

Sadly because of the optimized code, it was too hard for me to see why it got stuck in that loop. Removing the ping pong mode in our game, the problem has not come up again.

It's a bit disheartening to create a bug report with no repeatability at all. Sending the code to our game might not help either, as the bug only manifests once in a couple of hours of game play.

**Comment by tommo** [comment]
``` c++
void MOAITimer::DoStep ( float step ) {
    ...
    case PING_PONG: {
    ...
    while (( this->mTime < this->mStartTime ) || ( this->mTime >= this->mEndTime )) {

                ...
                    else {
                        this->mTime = this->mEndTime - ( this->mTime - this->mEndTime );
                ...
                    }
                ...
```

I believe when mTime == mEndTime the code above becomes a dead-loop.

**Comment by Kaali** [comment]
Confirmed, A simple fix could be to just break on equality:

``` cpp
        case PING_PONG: {

            if (( this->mTime < this->mStartTime ) || ( this->mTime >= this->mEndTime )) {

                while (( this->mTime < this->mStartTime ) || ( this->mTime >= this->mEndTime )) {

                    if ( this->mTime < this->mStartTime ) {
                        this->mTime = this->mStartTime + ( this->mStartTime - this->mTime );

                        float end = this->mTime < this->mEndTime ? this->mTime : this->mEndTime;
                        this->GenerateKeyframeCallbacks ( this->mStartTime, end, end == this->mEndTime );
                    }
                    else {
                        this->mTime = this->mEndTime - ( this->mTime - this->mEndTime );

                        float end = this->mTime > this->mStartTime ? this->mTime : this->mStartTime;
                        this->GenerateKeyframeCallbacks ( this->mEndTime, end, end == this->mStartTime );
                    }

                    this->mDirection *= -1.0f;
                    this->OnEndSpan ();
                    this->OnLoop ();
                    this->OnBeginSpan ();

                    if (this->mTime == this->mEndTime) break;
                }
            }
            else {
                this->GenerateKeyframeCallbacks ( t0, this->mTime, false );
            }
            result = this->mTime - t0;
            break;
        }
```

I think this way the callback logic would be consistent with other modes.

**Comment by CircuitCipher** [comment]
Fixed in PR #716
